findBean on missing bean broken in 3.0.x snapshots

Code:

```
return context.findBean(aClass, Qualifiers.byName(beanName)).isPresent();
```

Exception:

```
Caused by: io.micronaut.context.exceptions.NonUniqueBeanException: Multiple possible bean candidates found: [javax.inject.Provider, jakarta.inject.Provider, io.micronaut.context.BeanProvider]
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:2331)
	at io.micronaut.context.DefaultApplicationContext.findConcreteCandidate(DefaultApplicationContext.java:451)
	at io.micronaut.context.DefaultBeanContext.lastChanceResolve(DefaultBeanContext.java:3166)
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidateNoCache(DefaultBeanContext.java:3037)
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:2967)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2653)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1620)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1594)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:840)
	at io.micronaut.coherence.namespace.BeanBuilder.realize(BeanBuilder.java:66)
```

Related to the `@Any` stuff.

Will have a look on Monday

--- a/inject-java/src/test/groovy/io/micronaut/inject/any/AnyProviderSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/any/AnyProviderSpec.groovy
@@ -31,6 +31,8 @@ class AnyProviderSpec extends Specification {
 
         then:
         beanContext.getBean(Object, Qualifiers.byName("poodle")) instanceof Poodle
+        beanContext.findBean(Object, Qualifiers.byName("poodle")).isPresent()
+        !beanContext.findBean(Object, Qualifiers.byName("doesn't exist")).isPresent()
         beanContext.containsBean(Object, Qualifiers.byName("poodle"))
         !beanContext.containsBean(Object, Qualifiers.byName("doesnt-exist"))
         owner.dog instanceof Dog

--- a/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
+++ b/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
@@ -2322,14 +2322,7 @@ public class DefaultBeanContext implements BeanContext {
         if (qualifier instanceof AnyQualifier) {
             return candidates.iterator().next();
         } else {
-            final List<BeanDefinition<T>> withoutAnyBeans =
-                    candidates.stream().filter(bd -> !bd.hasDeclaredAnnotation(Any.class))
-                    .collect(Collectors.toList());
-            if (withoutAnyBeans.size() == 1) {
-                return withoutAnyBeans.iterator().next();
-            } else {
-                throw new NonUniqueBeanException(beanType, candidates.iterator());
-            }
+            throw new NonUniqueBeanException(beanType, candidates.iterator());
         }
     }
 
@@ -3667,7 +3660,6 @@ public class DefaultBeanContext implements BeanContext {
             filterReplacedBeans(null, candidates);
             Stream<BeanDefinition<T>> stream = candidates.stream();
             if (qualifier != null && !(qualifier instanceof AnyQualifier)) {
-                stream = stream.filter(bd -> !bd.hasDeclaredAnnotation(Any.class));
                 stream = qualifier.reduce(beanType.getType(), stream);
             }
             return stream.count() > 0;

--- a/inject/src/main/java/io/micronaut/inject/BeanDefinition.java
+++ b/inject/src/main/java/io/micronaut/inject/BeanDefinition.java
@@ -15,7 +15,6 @@
  */
 package io.micronaut.inject;
 
-import io.micronaut.context.annotation.Any;
 import io.micronaut.context.annotation.DefaultScope;
 import io.micronaut.context.annotation.Provided;
 import io.micronaut.core.annotation.AnnotationUtil;
@@ -98,54 +97,49 @@ public interface BeanDefinition<T> extends AnnotationMetadataDelegate, Named, Be
             return false;
         }
         if (BeanType.super.isCandidateBean(beanType)) {
-            if (hasStereotype(Any.class)) {
-                return true;
-            } else {
-                final Argument<?>[] typeArguments = beanType.getTypeParameters();
-                final int len = typeArguments.length;
-                Class<?> beanClass = beanType.getType();
-                if (len == 0) {
-                    if (isContainerType()) {
-                        final Optional<Argument<?>> containerElement = getContainerElement();
-                        if (containerElement.isPresent()) {
-                            final Class<?> t = containerElement.get().getType();
-                            return beanType.isAssignableFrom(t) || beanClass == t;
-                        } else {
-                            return false;
-                        }
+            final Argument<?>[] typeArguments = beanType.getTypeParameters();
+            final int len = typeArguments.length;
+            Class<?> beanClass = beanType.getType();
+            if (len == 0) {
+                if (isContainerType()) {
+                    final Optional<Argument<?>> containerElement = getContainerElement();
+                    if (containerElement.isPresent()) {
+                        final Class<?> t = containerElement.get().getType();
+                        return beanType.isAssignableFrom(t) || beanClass == t;
                     } else {
-                        return true;
+                        return false;
                     }
                 } else {
-                    final Argument<?>[] beanTypeParameters;
-                    if (!Iterable.class.isAssignableFrom(beanClass)) {
-                        final Optional<Argument<?>> containerElement = getContainerElement();
-                        //noinspection OptionalIsPresent
-                        if (containerElement.isPresent()) {
-                            beanTypeParameters = containerElement.get().getTypeParameters();
-                        } else {
-                            beanTypeParameters = getTypeArguments(beanClass).toArray(Argument.ZERO_ARGUMENTS);
-                        }
+                    return true;
+                }
+            } else {
+                final Argument<?>[] beanTypeParameters;
+                if (!Iterable.class.isAssignableFrom(beanClass)) {
+                    final Optional<Argument<?>> containerElement = getContainerElement();
+                    //noinspection OptionalIsPresent
+                    if (containerElement.isPresent()) {
+                        beanTypeParameters = containerElement.get().getTypeParameters();
                     } else {
                         beanTypeParameters = getTypeArguments(beanClass).toArray(Argument.ZERO_ARGUMENTS);
                     }
-                    if (len != beanTypeParameters.length) {
-                        return false;
-                    }
+                } else {
+                    beanTypeParameters = getTypeArguments(beanClass).toArray(Argument.ZERO_ARGUMENTS);
+                }
+                if (len != beanTypeParameters.length) {
+                    return false;
+                }
 
-                    for (int i = 0; i < beanTypeParameters.length; i++) {
-                        Argument<?> candidateParameter = beanTypeParameters[i];
-                        final Argument<?> requestedParameter = typeArguments[i];
-                        if (!requestedParameter.isAssignableFrom(candidateParameter.getType())) {
-                            if (!(candidateParameter.isTypeVariable() && candidateParameter.isAssignableFrom(requestedParameter.getType()))) {
-                                return false;
-                            }
+                for (int i = 0; i < beanTypeParameters.length; i++) {
+                    Argument<?> candidateParameter = beanTypeParameters[i];
+                    final Argument<?> requestedParameter = typeArguments[i];
+                    if (!requestedParameter.isAssignableFrom(candidateParameter.getType())) {
+                        if (!(candidateParameter.isTypeVariable() && candidateParameter.isAssignableFrom(requestedParameter.getType()))) {
+                            return false;
                         }
                     }
-                    return true;
                 }
+                return true;
             }
-
         }
         return false;
     }
@@ -366,7 +360,7 @@ public interface BeanDefinition<T> extends AnnotationMetadataDelegate, Named, Be
             return ReflectionUtils.EMPTY_CLASS_ARRAY;
         } else {
             final List<Argument<?>> typeArguments = getTypeArguments(type);
-            if (typeArguments.size() == 0) {
+            if (typeArguments.isEmpty()) {
                 return ReflectionUtils.EMPTY_CLASS_ARRAY;
             }
             Class[] params = new Class[typeArguments.size()];

--- a/inject/src/main/java/io/micronaut/inject/provider/AbstractProviderDefinition.java
+++ b/inject/src/main/java/io/micronaut/inject/provider/AbstractProviderDefinition.java
@@ -48,6 +48,7 @@ import java.util.Optional;
  */
 public abstract class AbstractProviderDefinition<T> implements BeanDefinition<T>, BeanFactory<T>, BeanDefinitionReference<T> {
 
+    private final static Argument<Object> TYPE_VARIABLE = Argument.ofTypeVariable(Object.class, "T");
     private final AnnotationMetadata annotationMetadata;
 
     public AbstractProviderDefinition() {
@@ -195,10 +196,8 @@ public abstract class AbstractProviderDefinition<T> implements BeanDefinition<T>
     @Override
     @NonNull
     public final List<Argument<?>> getTypeArguments() {
-        return Collections.singletonList(
-                Argument.OBJECT_ARGUMENT
-        );
-    }
+        return Collections.singletonList(TYPE_VARIABLE);
+    }    
 
     @Override
     public AnnotationMetadata getAnnotationMetadata() {

--- a/inject/src/main/java/io/micronaut/inject/provider/AbstractProviderDefinition.java
+++ b/inject/src/main/java/io/micronaut/inject/provider/AbstractProviderDefinition.java
@@ -48,7 +48,7 @@ import java.util.Optional;
  */
 public abstract class AbstractProviderDefinition<T> implements BeanDefinition<T>, BeanFactory<T>, BeanDefinitionReference<T> {
 
-    private final static Argument<Object> TYPE_VARIABLE = Argument.ofTypeVariable(Object.class, "T");
+    private static final Argument<Object> TYPE_VARIABLE = Argument.ofTypeVariable(Object.class, "T");
     private final AnnotationMetadata annotationMetadata;
 
     public AbstractProviderDefinition() {

--- a/inject/src/main/java/io/micronaut/inject/qualifiers/AnnotationMetadataQualifier.java
+++ b/inject/src/main/java/io/micronaut/inject/qualifiers/AnnotationMetadataQualifier.java
@@ -90,7 +90,7 @@ class AnnotationMetadataQualifier<T> extends NameQualifier<T> {
 
     @Override
     public <BT extends BeanType<T>> Stream<BT> reduce(Class<T> beanType, Stream<BT> candidates) {
-        if (annotationMetadata.hasDeclaredAnnotation(Any.class)) {
+        if (beanType != Object.class && annotationMetadata.hasDeclaredAnnotation(Any.class)) {
             return candidates;
         }
         String name;

--- a/inject/src/main/java/io/micronaut/inject/qualifiers/AnyQualifier.java
+++ b/inject/src/main/java/io/micronaut/inject/qualifiers/AnyQualifier.java
@@ -16,7 +16,6 @@
 package io.micronaut.inject.qualifiers;
 
 import io.micronaut.context.Qualifier;
-import io.micronaut.context.annotation.Any;
 import io.micronaut.core.annotation.Internal;
 import io.micronaut.inject.BeanType;
 
@@ -39,10 +38,7 @@ public final class AnyQualifier<T> implements Qualifier<T> {
 
     @Override
     public <BT extends BeanType<T>> Stream<BT> reduce(Class<T> beanType, Stream<BT> candidates) {
-        return candidates.filter(candidate -> {
-            Class<T> candidateType = candidate.getBeanType();
-            return true;
-        });
+        return candidates;
     }
 
     @Override

--- a/inject/src/main/java/io/micronaut/inject/qualifiers/AnyQualifier.java
+++ b/inject/src/main/java/io/micronaut/inject/qualifiers/AnyQualifier.java
@@ -16,6 +16,7 @@
 package io.micronaut.inject.qualifiers;
 
 import io.micronaut.context.Qualifier;
+import io.micronaut.context.annotation.Any;
 import io.micronaut.core.annotation.Internal;
 import io.micronaut.inject.BeanType;
 
@@ -31,14 +32,17 @@ import java.util.stream.Stream;
 @Internal
 public final class AnyQualifier<T> implements Qualifier<T> {
     @SuppressWarnings("rawtypes")
-    static final AnyQualifier INSTANCE = new AnyQualifier();
+    public static final AnyQualifier INSTANCE = new AnyQualifier();
 
     private AnyQualifier() {
     }
 
     @Override
     public <BT extends BeanType<T>> Stream<BT> reduce(Class<T> beanType, Stream<BT> candidates) {
-        return candidates;
+        return candidates.filter(candidate -> {
+            Class<T> candidateType = candidate.getBeanType();
+            return true;
+        });
     }
 
     @Override

--- a/inject/src/main/java/io/micronaut/inject/qualifiers/NameQualifier.java
+++ b/inject/src/main/java/io/micronaut/inject/qualifiers/NameQualifier.java
@@ -63,7 +63,7 @@ class NameQualifier<T> implements Qualifier<T>, io.micronaut.core.naming.Named {
             if (!beanType.isAssignableFrom(candidate.getBeanType())) {
                 return false;
             }
-            if (candidate.getAnnotationMetadata().hasDeclaredAnnotation(Any.class)) {
+            if (beanType != Object.class && candidate.getAnnotationMetadata().hasDeclaredAnnotation(Any.class)) {
                 return true;
             }
             String typeName;
@@ -124,7 +124,7 @@ class NameQualifier<T> implements Qualifier<T>, io.micronaut.core.naming.Named {
         return candidates.filter(candidate -> {
                 if (candidate.isPrimary() && Primary.class.getSimpleName().equals(annotationName)) {
                     return true;
-                } else if (candidate.getAnnotationMetadata().hasDeclaredAnnotation(Any.class)) {
+                } else if (beanType != Object.class && candidate.getAnnotationMetadata().hasDeclaredAnnotation(Any.class)) {
                     return true;
                 }
                 String candidateName;

